### ISSUE TITLE
Add canonical graph parameter binding v1 with atomic candidate application

### ISSUE THREAD

**Issue by Ahmadfauzi34** [Ahmadfauzi34/ai_agent_builder#163]
Follow-up to completed audit #161 / PR #162 (`94b5ae1ace294474f82017a188edb1b5f9ff4125`).

The audit proved that existing ES (`ask -> tell`) and `CompiledGraph` execution are individually sufficient, but the repository lacks a reusable, canonical boundary for mapping **one flat candidate vector** onto **all mutable trainable parameters referenced by a compiled graph**.

This is a generic graph-parameter boundary, not an ES-specific optimizer facade and not an autodiff feature.

## Problem proven by #162

Single-layer binding is natural through existing per-layer Registry APIs, but a multi-layer host currently has to:
1. inspect graph layer references;
2. choose an ordering convention;
3. deduplicate repeated references;
4. ask each layer for its weight layout/length;
5. invent offsets;
6. mutate each layer sequentially.

That final step is not atomic across layers: an earlier valid mutation can remain committed when a later layer slice fails. Manual rollback works as a host workaround, but must not become an implicit protocol.

## Production goal: graph parameter binding v1

Introduce a deterministic, optimizer-independent binding derived from a frozen `CompiledGraph` plan plus a currently valid `LayerRegistry` structural binding.

The binding should describe unique trainable parameter owners in deterministic **first-use graph-plan order**, reusing/extracting the same deduplication semantics already used by Program Bundle rather than introducing a competing ordering rule.

Each descriptor must bind at least:
- `layer_type`
- `layer_id`
- stable per-layer weight-layout identity/description
- flat candidate `offset`
- flat candidate `length`

Repeated references to one trainable layer appear once. Stateless layers consume zero coordinates. Unsupported referenced parameterized layer types must fail closed rather than be silently omitted.

## Required API semantics

Exact public naming may follow existing Rust conventions, but the boundary must provide the equivalents of:
- construct binding from `(CompiledGraph, LayerRegistry)`;
- inspect deterministic layout/descriptors and total parameter length;
- compute a binding identity/schema separate from `program-identity.v1`;
- flatten current bound parameters from the registry;
- validate an entire candidate against the binding without mutation;
- atomically apply one complete candidate.

### Identity
Existing structural `program-identity.v1` must remain unchanged by compatible mutable weight values.

Graph parameter binding therefore needs its own immutable identity, conceptually:

```text
binding_identity = H(
  schema
  || structural_program_identity
  || ordered(layer_type, layer_id, stable_weight_layout, offset, length)
)
```

Current weight **values** are excluded from binding identity.

Changing graph structure, referenced parameter owners, parameter lengths, or stable weight layout must change or invalidate the binding.

## Atomicity

The core invariant is:

```text
candidate invalid anywhere
    => registry unchanged everywhere
```

Do not implement the primary semantics as "mutate then rollback".

Prefer either:
- complete prevalidation of every slice and every setter failure condition, followed by a commit path that cannot fail; or
- stage compatible mutable layer state and commit only after the whole candidate is proven valid.

A malformed later slice must never leave an earlier layer changed.

### Deterministic layout
- `Linear -> ReLU -> Linear` produces deterministic first-use trainable owner order.
- repeated use of one Linear layer deduplicates to one owner.
- stateless ReLU consumes no candidate coordinates.
- identical graph + registry binding reproduces the same binding identity and descriptors.

### Structural invalidation
- compatible weight-value changes preserve both graph structural identity and binding identity.
- structural replacement / incompatible Registry binding is rejected.
- layout-affecting parameter structure changes require a new/recomputed binding.

### Flatten/apply round-trip
- flatten current graph parameters;
- apply a different valid complete vector;
- execution changes while graph structural identity remains stable;
- reapply original flat vector and reproduce original execution.

### Failure atomicity
- wrong total candidate length leaves every layer unchanged.
- malformed/non-finite candidate data, where the underlying layer contract rejects it, leaves every layer unchanged.
- construct pressure where an early layer slice is valid and a later layer would fail; prove zero mutation across all owners.

### Unsupported layer handling
- referenced trainable/parameterized layer types without a supported flat-weight contract fail closed.
- stateless layers are skipped explicitly, not by accidental error swallowing.

## Invariants / non-goals
- no autodiff, backprop, gradient tape, optimizer slots, or gradient API;
- no new optimizer and no change to OpenES / `(mu,lambda)` update formulas;
- no ES state inside `CompiledGraph` or `LayerRegistry`;
- no change to frozen graph plan format unless strictly necessary and separately justified;
- no reinterpretation of `program-bundle.v1` as a candidate format;
- no mutable weight values in existing structural program identity;
- no implicit host ordering convention;
- no Resolution/Authorization/proof/execution authority coupling;
- `Resolved != Authorized != Proven != Executed` remains intact.

## Staging
1. core Rust binding + direct tests;
2. only after core CI is fully green and merged, add a thin packaged-WASM adapter/artifact proof in a separate PR;
3. only after graph parameter binding v1 ships should an ES-to-graph objective/training facade be audited.

Autodiff remains intentionally deferred.
